src/buffer/base.ts

```diff
@@ -0,0 +1,447 @@
+   */
+  at(timestamp: number | bigint, unit: TimestampUnit = "us") {
+    if (!this.hasSymbols && !this.hasColumns) {
+      throw new Error(
+        "The row must have a symbol or column set before it is closed",
+      );
+    }
+    if (typeof timestamp !== "bigint" && !Number.isInteger(timestamp)) {
+      throw new Error(
+        `Designated timestamp must be an integer or BigInt, received ${timestamp}`,
+      );
+    }
+    const timestampNanos = timestampToNanos(BigInt(timestamp), unit);
+    const timestampStr = timestampNanos.toString();
+    this.checkCapacity([timestampStr], 2);
+    this.write(" ");
+    this.write(timestampStr);
+    this.write("\n");
+    this.startNewRow();
+  }
+
+  /**
+   * Closing the row without writing designated timestamp into the buffer. <br>
+   * Designated timestamp will be populated by the server on this record.
+   */
+  atNow() {
+    if (!this.hasSymbols && !this.hasColumns) {
+      throw new Error(
+        "The row must have a symbol or column set before it is closed",
+      );
+    }
+    this.checkCapacity([], 1);
+    this.write("\n");
+    this.startNewRow();
+  }
+
+  /**
+   * Returns the current position of the buffer. <br>
+   * New data will be written into the buffer starting from this position.
+   */
+  currentPosition(): number {
+    return this.position;
+  }
+
+  protected checkCapacity(data: string[], base = 0) {
+    let length = base;
+    for (const str of data) {
+      length += Buffer.byteLength(str, "utf8");
+    }
+    if (this.position + length > this.bufferSize) {
+      let newSize = this.bufferSize;
+      do {
+        newSize += this.bufferSize;
+      } while (this.position + length > newSize);
+      this.resize(newSize);
+    }
+  }
+
+  private compact() {
+    if (this.endOfLastRow > 0) {
+      this.buffer.copy(this.buffer, 0, this.endOfLastRow, this.position);
+      this.position = this.position - this.endOfLastRow;
+      this.endOfLastRow = 0;
+    }
+  }
+
+  protected writeColumn(
+    name: string,
+    value: unknown,
+    writeValue: () => void,
+    valueType?: string,
+  ) {
+    if (typeof name !== "string") {
+      throw new Error(`Column name must be a string, received ${typeof name}`);
+    }
+    if (valueType && typeof value !== valueType) {
+      throw new Error(
+        `Column value must be of type ${valueType}, received ${typeof value}`,
+      );
+    }
+    if (!this.hasTable) {
+      throw new Error("Column can be set only after table name is set");
+    }
+    this.checkCapacity([name], 2 + name.length);
+    this.write(this.hasColumns ? "," : " ");
+    validateColumnName(name, this.maxNameLength);
+    this.writeEscaped(name);
+    this.write("=");
+    writeValue();
+    this.hasColumns = true;
+  }
+
+  protected write(data: string) {
+    this.position += this.buffer.write(data, this.position);
+    if (this.position > this.bufferSize) {
```

Comment:
Since we reserve an extra byte, this check will work only for strings with ASCII chars. Not sure we should keep it or not.

Reply:
Removed it in https://github.com/questdb/nodejs-questdb-client/pull/49/commits/95d997f370aaad8f0fd7b54de9f4480f385b5fcb.
It is redundant because we extend the buffer when required.
This also means this check cannot be tested.
Instead, I added a test to assert that v2 buffer extends if a binary double does not fit in it.